keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,207 @@
+<!-- /toc -->
+
+## Summary
+
+Currently the identifier of the current actor (fieldManager) is set through an request option or the request user-agent,
+this KEP aims to provide a declarative way of setting the current fieldManager through a field in the object itself.
+
+## Motivation
+
+It is a common practice to use Kubernetes manifests as stored configuration of an application. Either to just introduce version control to the cluster's state, or to automate the process of updating objects.
+
+With the addition of Server Side Apply it could become necessary to change the way those manifests get applied, as the most benefit from this new feature requires every actor to define it's fieldManager. Doing this, currently is only possible through api options, the user-agent or as a kubectl option.
+
+To keep the way of interacting with manifests as easy as possible, this KEP suggests to allow defining the fieldManager as part of the manifest itself. Therefore no further action is required when applying a set of manifests and users can easily benefit from Server Side Apply without having to change their apply workflow.
+
+This would also assist existing solutions already interacting with manifest files, as the fieldManager information gets provided in a way they already know to handle which might improve compatibility.
+
+### Goals
+
+- Allow users to set their fieldManager information declaratively in their manifests
+
+### Non-Goals
+
+- Replace the existing ways to set the fieldManager information
+
+## Proposal
+
+To achieve the goals, we would add a field to ObjectMeta called `options`, this field would allow to send options to the api server.
+The field will be in the form of a map of string keys and values that gets validated to contain only known options (currently `fieldManager`).
+
+This field should not get persisted to storage and is solely for sending request options information to the apiserver.
+When the `options.fieldManager` value is not set or empty, the apiserver would fallback onto current behavior and default the fieldManager (or fail).
+
+This means, we add a field that is optional metadata, write-only and non-persisted and exists for the sole purpose of sending request information to the apiserver.
+The `options.fieldManager` value should accept a non empty string or be unset and should follow the same criteria as [ManagedFieldsEntry.Manager](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.14/#managedfieldsentry-v1-meta).
+
+Setting or not setting the `options.fieldManager` value will cause the following behavior for both apply and non-apply operations:
+
+- If the fieldManager request option is set for the request (for example through kubectl), it will get used as manager
+- If the newly introduced `options.fieldManager` option is set in the received request, it will get used as manager
+- If both of the two options are set and do not match, the request will get rejected.
```

Comment:
Isn't that 100% similar to what happens when you do `kubectl apply -n some-namespace -R -f some/resources/`?

Reply:
We've very considered this as an analogy and base use-case for this proposal.

Reply:
I also strongly suspect that this would become a kustomize flag that would propagate the option automatically to all configs, so hopefully this wouldn't be a problem.